tests/integration/workflows/test_workflows_tasks_actions.py

```diff
@@ -472,3 +479,79 @@ def test_count_reference_coreness(
         'core': 0,
         'non_core': 1,
     }
+
+
+@pytest.fixture
+def load_from_source_data_workflow(workflow_app):
+    def mess_up_data(obj, eng):
+        obj.data['titles'][0]['title'] += ' changed!'
+        obj.save()
+
+    class SourceDataWorkflow(object):
+        workflow = [load_from_source_data, mess_up_data]
+
+    workflow_app.extensions['invenio-workflows'].register_workflow(
+        'load_source_data',
+        SourceDataWorkflow,
+    )
+
+    yield SourceDataWorkflow
+
+    del workflow_app.extensions['invenio-workflows'].workflows['load_source_data']
+
+
+def test_workflow_loads_from_source_data_fails_on_no_source_data(
+    load_from_source_data_workflow,
+    workflow_app,
+    record_from_db,
+):
+    workflow_id = workflow_object_class.create(data_type='hep', data=record_from_db).id
+
+    with pytest.raises(ValueError) as exc:
+        start('load_source_data', object_id=workflow_id)
+
+    assert exc.match(r'source_data.*missing')
+
+
+def test_workflow_loads_from_source_data_preserves_task_history(
+    load_from_source_data_workflow,
+    workflow_app,
+    record_from_db,
+):
+    workflow_id = build_workflow(record_from_db).id
+
+    start('load_source_data', object_id=workflow_id)
+
+    workflow_object = workflow_object_class.get(workflow_id)
+
+    assert len(workflow_object.extra_data['_task_history']) == 1
+
+    workflow_object.callback_pos = [0]
+    workflow_object.save()
+    db.session.commit()
+
+    workflow_object.continue_workflow(start_point='restart_task', delayed=True)
+
+    assert len(workflow_object.extra_data['_task_history']) == 2
+
+
+def test_workflow_loads_from_source_data_restores_correctly(
+    load_from_source_data_workflow,
+    workflow_app,
+    record_from_db,
+):
+    workflow_id = build_workflow(record_from_db).id
+
+    start('load_source_data', object_id=workflow_id)
+
+    workflow_object = workflow_object_class.get(workflow_id)
+
+    assert workflow_object.data['titles'][0]['title'] == 'Fancy title for a new record changed!'
+
+    workflow_object.callback_pos = [0]
+    workflow_object.save()
+    db.session.commit()
```

Comment:
Remove the mess_up step, and manually modify the workflow, then restart it, and see that it got reverted.